src/ripple/app/tx/impl/DID.cpp

```diff
@@ -0,0 +1,215 @@
+        return temDISABLED;
+
+    if (ctx.tx.getFlags() & tfUniversalMask)
+        return temINVALID_FLAG;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    if (!ctx.tx.isFieldPresent(sfURI) && !ctx.tx.isFieldPresent(sfData))
+        return temEMPTY_DID;
+
+    if (ctx.tx.isFieldPresent(sfURI) && ctx.tx[sfURI].empty() &&
+        ctx.tx.isFieldPresent(sfData) && ctx.tx[sfData].empty())
+        return temEMPTY_DID;
+
+    return preflight2(ctx);
+}
+
+TER
+addSLE(
+    ApplyContext& ctx,
+    std::shared_ptr<SLE> const& sle,
+    AccountID const& owner)
+{
+    auto const sleAccount = ctx.view().peek(keylet::account(owner));
+    if (!sleAccount)
+        return tefINTERNAL;
+
+    // Check reserve availability for new object creation
+    {
+        auto const balance = STAmount((*sleAccount)[sfBalance]).xrp();
+        auto const reserve =
+            ctx.view().fees().accountReserve((*sleAccount)[sfOwnerCount] + 1);
+
+        if (balance < reserve)
+            return tecINSUFFICIENT_RESERVE;
+    }
+
+    // Add ledger object to ledger
+    ctx.view().insert(sle);
+
+    // Add ledger object to owner's page
+    {
+        auto page = ctx.view().dirInsert(
+            keylet::ownerDir(owner), sle->key(), describeOwnerDir(owner));
+        if (!page)
+            return tecDIR_FULL;
+        (*sle)[sfOwnerNode] = *page;
+    }
+    adjustOwnerCount(ctx.view(), sleAccount, 1, ctx.journal);
+    ctx.view().update(sleAccount);
+
+    return tesSUCCESS;
+}
+
+TER
+DIDSet::doApply()
+{
+    // Edit ledger object if it already exists
+    Keylet const didKeylet = keylet::did(account_);
+    if (auto const sleDID = ctx_.view().peek(didKeylet))
+    {
+        if (auto const uri = ctx_.tx[~sfURI])
+        {
+            if (uri->empty())
+            {
+                sleDID->makeFieldAbsent(sfURI);
+            }
+            else
+            {
+                (*sleDID)[sfURI] = *uri;
+            }
+        }
+        if (auto const data = ctx_.tx[~sfData])
+        {
+            if (data->empty())
+            {
+                sleDID->makeFieldAbsent(sfData);
+            }
+            else
+            {
+                (*sleDID)[sfData] = *data;
+            }
+        }
+        if (!sleDID->isFieldPresent(sfURI) && !sleDID->isFieldPresent(sfData))
+        {
+            return tecEMPTY_DID;
+        }
+        ctx_.view().update(sleDID);
+        return tesSUCCESS;
+    }
+
+    // Create new ledger object otherwise
+    auto const sleDID = std::make_shared<SLE>(didKeylet);
+    (*sleDID)[sfAccount] = account_;
+    if (auto const uri = ctx_.tx[~sfURI]; uri.has_value() && !uri->empty())
+        (*sleDID)[sfURI] = uri.value();
+    if (auto const data = ctx_.tx[~sfData]; data.has_value() && !data->empty())
+        (*sleDID)[sfData] = data.value();
+
+    if (auto const ret = addSLE(ctx_, sleDID, account_); !isTesSuccess(ret))
+        return ret;
+
+    return tesSUCCESS;
```

Comment:
The idea behind `addSLE` was that it could eventually become a helper function that all transactors can use - I noticed a lot of the transactors had the same code written, and this would make them (or at least future transactors) a bit easier to read.

Reply:
Simplified in https://github.com/XRPLF/rippled/pull/4636/commits/99b53beb144c66ed619563b94819f6a6deca8ca0